### ISSUE TITLE
👉👉👉👉 Help test the new GhostText version 👈👈👈👈 

### ISSUE THREAD

**Issue by fregante** [GhostText/GhostText#172]
### A brand-new version of GhostText is coming

The GhostText browser extension was rewritten from the ground up in #122  due to new restrictions introduced by browsers every so often.

**It hasn't been published yet** because it needs to be tested by you 💯 

0. Uninstall the current GhostText version from your browser
1. Download this zip file: [ghosttext-20.11.27.zip](https://github.com/GhostText/GhostText/files/5606349/ghosttext-20.11.27.zip)
2. Extract its contents into a folder
3. Follow these instructions to [load it in Chrome](https://www.smashingmagazine.com/2017/04/browser-extension-edge-chrome-firefox-opera-brave-vivaldi/#google-chrome-opera-vivaldi) or [Firefox](https://www.smashingmagazine.com/2017/04/browser-extension-edge-chrome-firefox-opera-brave-vivaldi/#mozilla-firefox).

## Issues?

Please include what you're testing:

- Browser?
- Editor?
- URL?
- Any issues that appear in the browser's console?
- Does the page work with the older version of GhostText?
- Does it work with Sublime Text 3?

## Main changes

- GhostText can edit multiple fields at once on the page, just click the GT button, select a field, click the button again, select another field, etc…
- To disconnect from the editor, either close the editor’s window or RIGHT CLICK the GT button.

**Comment by subnut** [comment]
@fregante 
There is no blue glowing border as there was before. Is this intended?

Firefox on Linux

**Comment by subnut** [comment]
## ISSUE: 
Even if the HTTP server is not currently running, GhostText extension shows _sucessfully connected_ and turns **ON** or, if already **ON**, the counter increases by one (indicating successful connection)

## EXPECTED:
Does not connect / shows error that server is not running

**Comment by subnut** [comment]
@fregante I forgot to mention in the above post that only the extension _icon_ indicates successful connection. 
The banner showing "Connected! You can switch to your editor" does **not** appear, nor does the console show "GhostText: Connected! You can switch to your editor" or "Sending -- characters"

**Comment by subnut** [comment]
## ISSUE: 
If a field is currently connected, then clicking the GhostText extension icon and selecting the connected field should disconnect it.
It currently does not disconnect.

## EXPECTED:
The selected field, if already connected, gets disconnected, but if not connected, then gets connected.

---
Firefox (latest stable release) on Arch Linux
GhostText test page [here](https://ghosttext.github.io/GhostText/demo/)

**Comment by subnut** [comment]
Ping @fregante

**Comment by fregante** [comment]
Thank you for the reports! I’m AFK for the next few days

**Comment by subnut** [comment]
> I’m AFK for the next few days 

@fregante How many days more? 😇

**Comment by subnut** [comment]
# Reddit doesn't work
(I don't know if the previous version worked! I never used it!)

## Explanation -
The text does appear when we are typing, but as soon as the browser window is focused, all the text is gone **from the browser window only.** The text **does not get deleted** in the editor itself

## See GIF below for demonstration -
![gifcast_201219003652](https://user-images.githubusercontent.com/56714626/102651525-b2756200-4192-11eb-9fd7-512b711fb743.gif)

---
Firefox on Arch Linux

**Comment by coltoneakins** [comment]
# Testing Report

This is a report testing the newer GhostText rewrite: ghosttext-20.11.27.zip

Context: https://github.com/GhostText/GhostText/issues/172

> Please include what you're testing:
> 
> Browser?
> Editor?
> URL?
> Any issues that appear in the browser's console?
> Does the page work with the older version of GhostText?
> Does it work with Sublime Text 3?


## Browser

Firefox 84.0

(Latest stable version on Ubuntu at the time of this writing.)

## Editor

Tested with both:

- Doom Emacs

- Sublime Text 3 (Build 3211)

## URL

Ace

https://ace.c9.io/

## Issues in Browser's Console?

**Yes, but the error is minor.**

The error may also just relate to the 3rd party plugin for Emacs. As in: this is probably not an error with the extension itself.

### Description of Issue

This isssue occurs when text is sent from editor -> browser in Emacs ONLY. This issue does not occur with Sublime Text 3. Browser -> editor is fine and produces no errors.

This error also does not affect the transmission of text. Both the editor and browser send updates to each other with no text loss.

Here is the error. It is sent everytime a character is typed in Emacs -> browser:

Error:

```
selections is undefined 5 content.js:150
```

Offending code in content.js line 150:

```
	receive(event) {
		const {
			text,
			selections
		} = JSON.parse(event.data);
		if (this.field.value !== text) {
			this.field.value = text;
		}

		this.field.selectionStart = selections[0].start; // <-- this is line 150
		this.field.selectionEnd = selections[0].end;
	}
```


![screenshot-ff-ace-doom-emacs-gt-rewrite](https://user-images.githubusercontent.com/16803020/102964450-4e63de00-44b1-11eb-974f-ccf266f97aba.png)

**I believe this is because the plugin for Emacs does not send any data on the current selection of text in the editor--it only sends the text itself.**


## Does the page work with the older version of GhostText?

**No, the page does not work.**
![screenshot-ff-ace-doom-emacs-gt-old](https://user-images.githubusercontent.com/16803020/102964448-4dcb4780-44b1-11eb-976d-bb60687b5c8a.png)

The page produces the error mentioned in this bug: https://github.com/GhostText/GhostText/issues/104#issue-263692210


This new error mentioned in the report also persists with the old version. However, I think it is because of the implemenetation of the 3rd party plugin for Emacs.


## Does it work w/ Sublime Text 3?

Yes, no errors.

![screenshot-ff-ace-sublime-text-gt-rewrite](https://user-images.githubusercontent.com/16803020/102964451-4e63de00-44b1-11eb-8a4f-ee0006c8fe3a.png)


---
---


**TL;DR**

We can probably close this issue: https://github.com/GhostText/GhostText/issues/104


Thank you Federico!

**Comment by subnut** [comment]
Ping @fregante for updates.....

<br>

---

Happy New Year to everybody :tada:

**Comment by fregante** [comment]
Thank you for the many reports! It looks like a lot of parts are still broken so it needs further work. The update was even rejected by Google so that confirms it 😬

**Comment by subnut** [comment]
>  The update was even rejected by Google so that confirms it

I didn't understand what you meant by that...

**Comment by kujohn** [comment]
Use Firefox and latest version of [atomic chrome](https://melpa.org/#/atomic-chrome) [editor.p5js.org](http://editor.p5js.org).
![2021-01-08 19 20 18](https://user-images.githubusercontent.com/1356936/104081917-cc7cf000-51e6-11eb-871a-6cdca603be00.gif)

**Comment by luisherranz** [comment]
Working great here with last Chrome and VimR. I love it! 😄

I've made a couple of PRs (https://github.com/GhostText/GhostText/pull/176 and https://github.com/GhostText/GhostText/pull/175). One to fix the Chrome focus after disconnection and another to not to ask for a textarea if there's an already selected one.

By the way, it works great with Slack messages! (the previous one didn't).

**Comment by fregante** [comment]
I opened a PR with some updates. Bear with me

> Reddit (markdown mode) doesn't work

@subnut It's possible that Reddit still considers the field "untouched" becuse GhostText didn't send any events, so it blindly resets it. I resolved https://github.com/GhostText/GhostText/issues/42 so this should be 

> **I believe this is because the plugin for Emacs does not send any data on the current selection of text in the editor--it only sends the text itself.**

@coltoneakins Indeed! I considered making the error silent, but since `selections` are part of the API, it should be fixed in the Emacs plugin (which is by third party, so you'll have to open an issue there). I think this is unrelated to Ace editor.

> > The update was even rejected by Google so that confirms it
> 
> I didn't understand what you meant by that...

@subnut for some reason the Chrome team completely rejected the update, but then I tried resubmitting the old version of GhostText and they accepted it (so it's not an issue with the _type_ of extension). I assume they rejected it because they couldn't get it to work.

**Comment by fregante** [comment]
I'll try publishing the current version. Let's hope they don't reject it, because currently Firefox rejected the 19.x version (#177) and Chrome rejected this 20.x version (https://github.com/GhostText/GhostText/issues/172#issuecomment-755922046)

**Comment by fregante** [comment]
Version 21.1.1 has been approved by the Chrome store so you should receive the update soon! 👍 

Firefox not yet
